sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -652,6 +653,299 @@ case class MapObjects private(
+    val tupleLoopVar = s"CollectObjectsToMap_tupleLoopValue$id"
+    val builderValue = s"CollectObjectsToMap_builderValue$id"
+    CollectObjectsToMap(
+      keyLoopValue, keyLoopIsNull, keyElementType, keyFunction(keyLoopVar), keyInputData,
+      valueLoopValue, valueLoopIsNull, valueElementType, valueFunction(valueLoopVar),
+      valueInputData,
+      tupleLoopVar, collClass, builderValue)
+  }
+}
+
+/**
+ * An equivalent to the [[MapObjects]] case class but returning an ObjectType containing
+ * a Scala collection constructed using the associated builder, obtained by calling `newBuilder`
+ * on the collection's companion object.
+ *
+ * @param keyLoopValue the name of the loop variable that is used when iterating over the key
+ *                     collection, and which is used as input for the `keyLambdaFunction`
+ * @param keyLoopIsNull the nullability of the loop variable that is used when iterating over
+ *                      the key collection, and which is used as input for the `keyLambdaFunction`
+ * @param keyLoopVarDataType the data type of the loop variable that is used when iterating over
+ *                           the key collection, and which is used as input for the
+ *                           `keyLambdaFunction`
+ * @param keyLambdaFunction A function that takes the `keyLoopVar` as input, and is used as
+ *                          a lambda function to handle collection elements.
+ * @param keyInputData An expression that when evaluated returns a collection object.
+ * @param valueLoopValue the name of the loop variable that is used when iterating over the value
+ *                       collection, and which is used as input for the `valueLambdaFunction`
+ * @param valueLoopIsNull the nullability of the loop variable that is used when iterating over
+ *                        the value collection, and which is used as input for the
+ *                        `valueLambdaFunction`
+ * @param valueLoopVarDataType the data type of the loop variable that is used when iterating over
+ *                             the value collection, and which is used as input for the
+ *                             `valueLambdaFunction`
+ * @param valueLambdaFunction A function that takes the `valueLoopVar` as input, and is used as
+ *                            a lambda function to handle collection elements.
+ * @param valueInputData An expression that when evaluated returns a collection object.
+ * @param tupleLoopValue the name of the loop variable that holds the tuple to be added to the
+  *                      resulting map (used only for Scala Map)
+ * @param collClass The type of the resulting collection.
+ * @param builderValue The name of the builder variable used to construct the resulting collection.
+ */
+case class CollectObjectsToMap private(
+    keyLoopValue: String,
+    keyLoopIsNull: String,
+    keyLoopVarDataType: DataType,
```

Comment:
I modelled this class after the `MapObjects` class so that they could be used similarly. I noticed that since then a new `UnresolvedMapObjects` class was introduced which also doesn't require the element data type. Would this be something similar? And if so, shouldn't I rather introduce a new `UnresolvedCollectObjectsToMap` class instead?